Implement caching

sergesemashko opened the pull request:
<!---
Please read the following. Pull requests that do not adhere to these guidelines will be closed.

Each pull request must, with the exception of minor documentation fixes, be associated with an open issue. If a corresponding issue does not exist please stop. Instead, create an issue so we can discuss the change first.

If there is an associated open issue, then the next step is to make sure you've read the relevant developer guide:

- Adding a rule: https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#write-the-rule

- Adding an option: https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-options-to-existing-rules

- Fixing a bug: https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-bugs

Once you've done that, then please continue by answering these two questions:  -->

> Which issue, if any, is this issue related to?

https://github.com/stylelint/stylelint/issues/2270

> Is there anything in the PR that needs further explanation?

UPDATED:
Tests are ~~in progress~~ **done**.
~~I decided to submit a PR for initial feedback before going too deep into woods with tests.~~

New dependencies:
- [debug](https://www.npmjs.com/package/debug) - for easy debugging/execution time profiling. To profile lint time just prefix you stylelint command with `DEBUG=stylelint:standalone stylelint "*.scss"`
- [imurmurhash](https://www.npmjs.com/package/imurmurhash) - see [eslint requires nodejs built with ssl support #5522](https://github.com/eslint/eslint/issues/5522)
- [file-entry-cache](https://www.npmjs.com/package/file-entry-cache) - required to get and cache file metadata
- [fs-promise](https://www.npmjs.com/package/fs-promise) - a promisified `fs-extra` version to read, copy, remove files using promises in tests.

Accounted for issues ESlint had with implementing cache:

- [Fix: invalidate cache when config changes. (fixes #3770)](https://github.com/eslint/eslint/pull/3818)
- [Update: --cache-file can take a directory where to store the cache. Fixes #4241](https://github.com/eslint/eslint/pull/4255)
- [Fix: `cacheLocation` handles paths in windows style. (fixes #4285)](https://github.com/eslint/eslint/pull/4287)
- [New: do not remove non visited files from cache. (Fixes #6780)](https://github.com/eslint/eslint/pull/6921)
- [eslint requires nodejs built with ssl support #5522](https://github.com/eslint/eslint/issues/5522)

@jeddy3, @davidtheclark, please, take a look

evilebottnawi commented:
@sergesemashko Awesome, thanks for helping :+1:

IanVS commented:
Looks like you did a great job researching the lessons we learned over at ESLint.  I'm very excited to see this functionality land in stylelint.  👏

sergesemashko commented:
I have more free time this week to submit tests as well, I'm close

sergesemashko commented:
Travis and appveyor are running out of memory in the very beginning, not sure why :/

On my local all tests are passing on node 7.5.0:
![](https://api.monosnap.com/rpc/file/download?id=IRRWbtTG30mMgYSwcBRIWV8gPcGwlW)

@ntwb, @evilebottnawi @davidtheclark, any ideas or suggestions?

m-allanson commented:
@sergesemashko this is an exciting feature, thanks for working on it!

When we first started using Jest we had some issues with CI failing due running out of memory. It looks like that is what's causing the failures on this PR too.

I've tried running your branch locally and noticed that the tests take significantly longer to run when compared to `stylelint:master`. There are also random failing tests due to async callback timeouts.

Here's the tests on `stylelint:master` running on my machine:

<img width="371" alt="screen shot 2017-02-07 at 22 57 57" src="https://cloud.githubusercontent.com/assets/381801/22716138/5eb7721a-ed8c-11e6-9717-3054822d4172.png">

Here's the tests on `sergesemashko:implement-caching` running on my machine:

<img width="756" alt="screen shot 2017-02-07 at 23 05 53" src="https://cloud.githubusercontent.com/assets/381801/22716150/6fa42dde-ed8c-11e6-9ea3-65a8b571394f.png">

(Note that the failing tests will be different on each run)

The `stylelint:master` tests run in around ~90 seconds, and the `implement-caching` tests around ~680 seconds. This is a big jump and I suspect it's why the CI is failing. I haven't taken a close look at the code but hopefully that's enough info someone to do some investigation.

sergesemashko commented:
@m-allanson, thanks for checking and pointing out to that. I'll investigate the reason. Definitely don't want to introduce any issues for testing :)

davidtheclark commented on docs/user-guide/node-api.md:
```diff
@@ -77,6 +77,20 @@ If `true`, all disable comments (e.g. `/* stylelint-disable block-no-empty */`)
+
+Store the info about processed files in order to only operate on the changed ones. The cache is stored in `.stylelintcache` by default. Enabling this option can dramatically improve Stylelint's running time by ensuring that only changed files are linted.
+
+**Note:** If you run Stylelint with `--cache` and then run Stylelint without `--cache`, the `.stylelintcache` file will be deleted. This is necessary because the results of the lint might change and make `.stylelintcache` invalid. If you want to control when the cache file is deleted, then use `--cache-location` to specify an alternate location for the cache file.
```
This is a little confusing to me: does that mean that if I use `--cache-location`, my cache file will become outdated when I run without `--cache`, because it will *not* be deleted? I'm not sure I'd want this implicit behavior change solely because I specified a cache file location.

sergesemashko replied:
Actually, cache file will be deleted when `--cache-location` is not changing between runs with `--cache` and without. Cache file won't be deleted for the following case:
1. `stylelint *.scss --cache` - create cache file in default location
1. `stylelint *.scss --cache-location /tmp` - won't delete cache file from default location
1. `stylelint *.scss --cache` - reuse cache file from default location

davidtheclark replied:
Ok, thanks for clarifying. I think we should remove this sentence from the docs because it represents an edge-case that is fairly confusing. If people ask about it, we could add a note somewhere else, like the FAQ.

davidtheclark commented on docs/user-guide/node-api.md:
```diff
@@ -77,6 +77,20 @@ If `true`, all disable comments (e.g. `/* stylelint-disable block-no-empty */`)
+
+Path to the cache location. Can be a file or a directory. If no location is specified, `.stylelintcache` will be used. In that case, the file will be created in the directory where the `stylelint` command is executed.
+
+If a directory is specified, a cache file will be created inside the specified folder. The name of the file will be based on the hash of the current working directory (CWD). e.g.: `.cache_hashOfCWD`
```
Why use a hash filename instead of simply adding `.stylelintcache` to that directory?

sergesemashko replied:
This will allow to reuse a single location for the caches from different projects and still receive the benefits of the cache. I'll add this sentence to the doc.
see https://github.com/eslint/eslint/pull/4255#issue-113211086.

davidtheclark replied:
Sounds good!

davidtheclark commented on lib/__tests__/fixtures/cache/valid.css:
```diff
@@ -0,0 +1 @@
+/* This will not cause a CSS syntax error */
```
Do you mean this will not cause a *linting* error? Those "invalid" files also will not cause CSS syntax errors, right?

sergesemashko replied:
yep, right, I'll update the wording

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,73 @@
+      mockedFileCache.verify()
+    })
+  })
+  afterEach(() => {
```
Can you please put this up by the `beforeEach`? I find that helps force me to grasp the full context of each test upfront.

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,73 @@
+      expect(destroyStub.called).toBe(true)
+    })
+  })
+  afterEach(() => {
```
Same point about moving this up.

davidtheclark commented on lib/standalone.js:
```diff
@@ -76,14 +91,28 @@ module.exports = function (options/*: Object */)/*: Promise<stylelint$standalone
       }).then(postcssResult => {
-        return stylelint._createStylelintResult(postcssResult, filePath)
+        if (postcssResult.stylelint.stylelintError && useCache) {
+          debug(`File has problems, skip caching: ${absoluteFilepath}`)
```
Let's be more clear with this message. How about `${absoluteFilepath} contains linting errors and will not be cached.`?

davidtheclark commented on lib/standalone.js:
```diff
@@ -100,6 +129,10 @@ module.exports = function (options/*: Object */)/*: Promise<stylelint$standalone
+    if (useCache) {
+      fileCache.reconcile()
+    }
+    debug(`Linting complete in: ${Date.now() - startTime}ms`)
```
No need for a colon here.

davidtheclark commented on lib/utils/FileCache.js:
```diff
@@ -0,0 +1,48 @@
+const path = require("path")
+const debug = require("debug")("stylelint:file-cache")
+const getCacheFile = require("./getCacheFile")
+const DEFAULT_CACHE_LOCATION = "./.stylelintcache"
```
Let's add an empty line above this constant to differentiate it from the dependencies.

davidtheclark commented on lib/utils/FileCache.js:
```diff
@@ -0,0 +1,48 @@
+  const changed = descriptor.changed || meta.hashOfConfig !== this._hashOfConfig
+  if (!changed) {
+    // since the file passed we store the result here
+    // successful runs as it will always should be 0 error 0 warnings
```
This comment is unclear — could you please try rephrasing?

davidtheclark commented on lib/utils/getCacheFile.js:
```diff
@@ -0,0 +1,74 @@
+ * a directory or looks like a directory (ends in `path.sep`), in which case the file
+ * name will be the `cacheFile/.cache_hashOfCWD`
+ *
+ * if cacheFile points to a file or looks like a file then in will just use that file
```
I'd like to see capitalization and punctuation in all of these comments. If English is not your first language, feel free to ask for help and someone can pitch in.

sergesemashko replied:
Sorry, didn't read through carefully copied text

davidtheclark commented on lib/utils/getCacheFile.js:
```diff
@@ -0,0 +1,74 @@
+  cacheFile = path.normalize(cacheFile)
+
+  const resolvedCacheFile = path.resolve(cwd, cacheFile)
+  const looksLikeADirectory = cacheFile[cacheFile.length - 1 ] === path.sep
```
Extra space before the `]`.

davidtheclark commented on lib/utils/getCacheFile.js:
```diff
@@ -0,0 +1,74 @@
+    return getCacheFileForDirectory()
+  }
+
+  return resolvedCacheFile
```
Could line 48-73 be simplified to this?

```js
if (looksLikeADirectory || (fileStats && fileStats.isDirectory()) {
  return getCacheFileForDirectory()
}

return resolvedCacheFile
```

sergesemashko replied:
makes sense

davidtheclark commented on lib/utils/getCacheFile.js:
```diff
@@ -0,0 +1,74 @@
+ * if cacheFile points to a file or looks like a file then in will just use that file
+ *
+ * @param {string} cacheFile The name of file to be used to store the cache
+ * @param {string} cwd Current working directory
```
Let's note that this is only useful for testing.

davidtheclark commented on lib/utils/__tests__/FileCache.test.js:
```diff
@@ -0,0 +1,119 @@
+  expect(fileCache._hashOfConfig).toBe(hashOfConfig)
+  mockedFileEntryCache.verify()
+  // restore mocked objects
+  mockedFileEntryCache.restore()
```
What is the purpose of these two lines? Am I missing an assertion that they relate to?

sergesemashko replied:
agree, line 23 doesn't make sense. Line 24 restores file-entry-cache dependency for further use in case of new tests.

davidtheclark commented on lib/utils/__tests__/FileCache.test.js:
```diff
@@ -0,0 +1,119 @@
+  mockedFileEntryCache.restore()
+  mock.stopAll()
+})
+it("reconcile() stores hash to descriptor", () => {
```
I'm not sure we need this test. Shouldn't we be able to run tests that determine whether `FileCache` actually *works* as expected without caring about the implementation details? This seems like it's just testing implementation details, not the API of the `FileCache`.

sergesemashko replied:
FileCache.prototype.reconcile() wraps `file-cache-entry`.reconcile() to add config hash before saving to every file entry. Just trying to make sure config hash is added on reconcile

davidtheclark commented on lib/utils/__tests__/FileCache.test.js:
```diff
@@ -0,0 +1,119 @@
+      getFileDescriptor: getFileDescriptorStub,
+    },
+    _hashOfConfig: hashOfConfig,
+  }
```
Instead of creating this mock `fileCache` object, can't you use `stub` to `stub` the *real* `getFileDescriptor` method on a real `FileCache` instance? If so, wouldn't that would also avoid the need to awkwardly use `call` below?

sergesemashko replied:
ok, since I'm going to have more real things in tests, that would be the new way of testing hasFileChanged

davidtheclark commented on lib/utils/__tests__/FileCache.test.js:
```diff
@@ -0,0 +1,119 @@
+  expect(hasFileChanged.call(fileCache, invalidFile1)).toBe(true)
+  expect(hasFileChanged.call(fileCache, invalidFile2)).toBe(true)
+})
+it("file-cache-entry methods are called", () => {
```
Why does this matter for our tests? Aren't these implementation details?

sergesemashko replied:
Just was trying to make sure the constructor is covered by tests. Maybe doesn't make sense when real FileCache is used elsewhere, I'll if it's still needed later

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,73 @@
+const fixturesPath = path.join(__dirname, "fixtures")
+
+describe("standalone cache is enabled", () => {
+  let mockedFileCache
```
Why are we mocking the file cache instead of testing the real thing? Couldn't we actually *use* the file cache in our tests, cleaning up after as needed, and then bypass some of the efforts below to test its internals?

sergesemashko replied:
Makes sense, I'll change to the real thing

davidtheclark commented:
Awesome work @sergesemashko!

sergesemashko commented:
What's in the latest update: 
- All the tests are rewritten to test on real dependencies. Actually, tests replicate real use cases now. The only thing I'm not sure about whether AppVeyor allows to do file manipulations in tests, because tests are failing there, when everything is ok on local. @evilebottnawi, @davidtheclark, do you have any insights on it or preferences how to deal with file manipulations in tests?
- Reflected comments in the code, @davidtheclark thanks a bunch!
- Reduced the overhead by adding caching and execution time looks great now.

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,146 @@
+const standalone = require("../standalone")
+const hash = require("../utils/hash")
+const fixturesPath = path.join(__dirname, "fixtures")
+const fsExtra = require("fs-extra")
```
I'd rather we avoid this new dependency and just use regular `fs`. For copying you could use https://github.com/sindresorhus/cpy. Additionally, we should be able to avoid sync calls for all of this because Jest has good async/Promise support.

sergesemashko replied:
thanks, wasn't aware of async/promise support

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,146 @@
+      expect(output.errored).toBe(true)
+      // Ensure only changed files are linted
+      const isValidFileLinted = !!output.results.find(file => file.source === validFile)
+      const isInvalidFileLinted = !!output.results.find(file => file.source === changedFile)
```
Should that be `invalidFile` instead of `changedFile`?

sergesemashko replied:
I realized that changedFile is a confusing var name here, it should be something like `newFileDest`. The idea here is to create new file so standalone should process both cached and uncached file.

davidtheclark commented on lib/__tests__/standalone-cache.test.js:
```diff
@@ -0,0 +1,146 @@
+      // Ensure cache file doesn't contain invalid file
+      const cachedFiles = fsExtra.readJsonSync(expectedCacheFilePath)
+      expect(typeof cachedFiles[validFile] === "object").toBe(true)
+      expect(typeof cachedFiles[changedFile] === "undefined").toBe(true)
```
Again, wondering if this should be `invalidFile`.

sergesemashko replied:
I'll update the var as per [comment](https://github.com/stylelint/stylelint/pull/2293#discussion_r102703828)

davidtheclark commented:
@jeddy3 I think we're close on this one. Do you want to give it a review? It's a pretty significant feature, so I want to make sure I'm not the only one who looks it over.

sergesemashko commented:
@davidtheclark, thanks for the copy edits and feedback. 
Maybe you have any ideas why file operations are failing in appveyor build?

davidtheclark commented:
@sergesemashko I'm afraid not. I try to ignore Appveyor. We have some Windows users in @stylelint/core. One guess: have you used `path` functions to normalize path separators?

sergesemashko commented:
@davidtheclark, good idea, I'll try normalized paths in tests.

sergesemashko commented:
@davidtheclark, Here are the latest updates:
1. Sync calls are converted to promises, however I decided to leave `fs-promise`(a promisified `fs-extra` version) because It required ways to read, remove and copy files using promises. And I though implementing the same functions using just `fs` would be redundant.
2. I updated all paths to use normalized paths, however the issue was completly gone when I added normalize to paths returned by globby()

sergesemashko commented:
@davidtheclark @evilebottnawi @jeddy3, could you please take a look?

evilebottnawi commented:
@sergesemashko can you squash commits for easy review

sergesemashko commented:
@evilebottnawi, sorry for the delay, It was tricky to squash all the commits in a nice way because there were merges from master. So, I decided to apply all the changes on a clean master branch within one commit.

cc/ @davidtheclark @jeddy3 

Test time looks good now, BTW:
![](https://api.monosnap.com/rpc/file/download?id=FydCoXhrW6Vd8tVOPPS6CVulJk90jl)

davidtheclark commented:
@sergesemashko Sorry for the delay on this ... I've had an exceptionally busy few weeks. I'm hoping I can review and merge the PR this week.

Related question: Would you like to become part of the stylelint org? This is a very significant feature contribution, and we'd certainly love to have you on board if it's something you're interested in.

sergesemashko commented:
@davidtheclark, thank you. Certainly yes, I'd love to become a part of stylelint org.

evilebottnawi commented:
@sergesemashko Some question:
1. What about invalidate cache file if i change configuration?
2. Why we use multiple `cache` files? We can use one cache file and meta information about cache as do `eslint`.
Example
```shell
{"/absolute/path/to/index.js":{"size":81,"mtime":1475500756217,"hashOfConfig":"114krsx","results":{"filePath":"/absolute/path/to/index.js","messages":[],"errorCount":0,"warningCount":0}},
```
3. We can use `pify` instead `fs-promise` https://github.com/sindresorhus/pify. More universal package for promising callbacks.
4. I think we should remove all debug function from this commit and do this in separate PR.

sergesemashko commented:
@evilebottnawi, thanks for your feedback.

> - What about invalidate cache file if i change configuration?

it's already accounted. Every cached entry has the hash of the config. If config changes all the files will be re-linted because hashes don't match. Checkout [this line](https://github.com/stylelint/stylelint/pull/2293/files#diff-9de7ea0960deccf5a8e2c5eaadc6c91aR24)
> - Why we use multiple cache files? 556168c#diff-25bbf797b85423b576b2d765600b119aR33

This will allow to reuse a single location (let's say `/tmp`) for the caches from different projects and still receive the benefits of the cache. I'll add this sentence to the doc.
see eslint/eslint#4255 (comment).
> - We can use pify instead fs-promise https://github.com/sindresorhus/pify. More universal package for promising callbacks.

Thanks, wasn't aware of this package. Looks like it may be helpful for any other cases and it looks much lighter that `fs-promise`. Will update soon.
> - I think we should remove all debug function from this commit and do this in separate PR. 

Didn't quite get your concerns. Do you want to add more debug messages in separate PR? I'd like keep it here, if possible, as profiling time rely on using `debug` and `debug()` usages look pretty straightforward. Separate PR would just mean more work for me and for the reviewers.

evilebottnawi commented:
@sergesemashko Thanks for answer.

> Didn't quite get your concerns. Do you see potential issues with introducing debug and want keep discussion in separate PR?
TBH, I'd like keep it here, if possible, as cache profiling rely on using debug and its usages look pretty straightforward.
Separate PR would just mean more work for me and for the reviewers. If you know something that might be affected by adding debug within this PR, please let me know.

No i don't see problems, we should add `debug` for all code or nothing. If this is already done, it will be fine :smile: 

SGTM for me :1st_place_medal:

sergesemashko commented:
@evilebottnawi , I see. Frankly speaking, adding debug for all codebase sounds like a "big deal" for me 😃 I'm not yet that familiar with all the cases where it really should be added.
Maybe, gradually introducing debug to the file as we go would work? 
Debug namespaces are already unique per file, e.g. `stylelint:standalone`, `stylelint:file-cache`, so it can be easily introduced to other files when necessary. 
I know it sounds slippery, but I think it's worth suggesting :) If it still doesn't work, I'll extract `debug` and will keep the discussion in separate PR. 
cc/ @davidtheclark

davidtheclark commented:
@jeddy3 are you good with merging this?

jeddy3 commented:
Added to changelog:

> -   Added: `cache` option to store the info about processed files in order to only operate on the changed ones the next time you run stylelint ([#2293](https://github.com/stylelint/stylelint/pull/2293)).

evilebottnawi commented:
@sergesemashko I experienced the work of the cache and it's great! Only I noticed one strangeness:
we ignore by default `node_modules` and `bower_components`, but if package have .scss (or .css) and i update package, this files fall as if not cached. I think problems in here https://github.com/stylelint/stylelint/blob/master/lib/standalone.js#L144 you don't exclude files from `node_modules` and `bower_components`.

My npm script: `stylelint '{**/*,*}.scss' --ignore-path .gitignore`. I don't know it is bug or not, just information.

sergesemashko commented:
@evilebottnawi, thanks for the input. 
Cache info for specific file is used depending on the output of `file-entry-cache` that checks if file has changed. And cache is invalidated for all changed files. 
Could you please clarify what's your expectation from linting dependency styles?

evilebottnawi commented:
@sergesemashko default command `stylelint '{**/*,*}.scss' --ignore-path .gitignore` globbing all files also inside in node_modules :(

olegskl commented:
I was wondering why is `fileCache` [initialized on line 105](https://github.com/stylelint/stylelint/blob/43b18fe4c56e9f177644b50354fb8efe9fb34a84/lib/standalone.js#L105), but `prepareReturnValue` [is called on line 90](https://github.com/stylelint/stylelint/blob/43b18fe4c56e9f177644b50354fb8efe9fb34a84/lib/standalone.js#L90) inside the return statement? When there are no files (linting code as a string) this throws:

> Cannot read property 'reconcile' of undefined

evilebottnawi commented:
@olegskl can you create issue?

sergesemashko commented:
@olegskl, thanks for reporting this. fileCache is initialized after 
```js
if (!files) {
    const absoluteCodeFilename = (codeFilename !== undefined && !path.isAbsolute(codeFilename))
      ? path.join(process.cwd(), codeFilename)
      : codeFilename
    return stylelint._lintSource({
      code,
      codeFilename: absoluteCodeFilename,
    }).then(postcssResult => {
      return stylelint._createStylelintResult(postcssResult)
    }).catch(handleError).then(stylelintResult => {
      return prepareReturnValue([stylelintResult])
    })
  }
```
because this block is executed when raw styles are passed. So, there is no need to initialize `fileCache` before this block, because there is no files to cache. 

However, looks like you found a bug, which wasn't covered by tests. @olegskl, could you please submit an issue? And I'll take a look at it.

olegskl commented:
@sergesemashko here it is #2492, sorry for the late reply.
